qupath-extension-bioformats/src/main/java/qupath/lib/images/servers/bioformats/BioFormatsImageServer.java

```diff
@@ -1376,16 +1376,24 @@ BufferedImage openImage(TileRequest tileRequest, int series, int nChannels, bool
 				if (tileWidth <= 0 || tileHeight <= 0) {
 					throw new IOException("Unable to request pixels for region with downsampled size " + tileWidth + " x " + tileHeight);
 				}
-
-				// OME Zarr files provide z scaling (the number of z stacks decreases when the resolution becomes
-				// lower, like the width and height)
-				if (ipReader instanceof ZarrReader) {
-					z /= tileRequest.getDownsample();
-				}
 		
 				synchronized(ipReader) {
 					ipReader.setSeries(series);
 					ipReader.setResolution(level);
+
+					// Some files provide z scaling (the number of z stacks decreases when the resolution becomes
+					// lower, like the width and height), so z needs to be updated for levels > 0
+					if (level > 0) {
```

Comment:
I don't know if `ipReader.setResolution` is expensive. In case it is, we should try to avoid calling it.
* Here, I think you can put this block *before* the initial `ipReader.setResolution(level)` and include that in an `else` (since this block set the resolution to the level at the end anyway).
* You can also use `if (level > 0 && z > 0)` since otherwise it shouldn't matter